client/src/main/java/io/pravega/client/state/impl/ReaderGroupStateSynchronizer.java

```diff
@@ -0,0 +1,137 @@
+import static io.pravega.shared.NameUtils.getStreamForReaderGroup;
+
+/**
+ * A new {@link StateSynchronizer} implementation specifically with the {@link ReaderGroupState} to allow for special
+ * updates to the controller.
+ */
+public class ReaderGroupStateSynchronizer implements StateSynchronizer<ReaderGroupState> {
+    private final String readerGroup;
+    private final StateSynchronizer<ReaderGroupState> synchronizer;
+    private final Controller controller;
+
+    /**
+     * Creates a new instance of ReaderGroupStateSynchronizer class.
+     *
+     * @param readerGroup Name of the {@link io.pravega.client.stream.ReaderGroup}.
+     * @param synchronizer The {@link StateSynchronizer} instance with {@link ReaderGroupState}.
+     * @param controller The {@link Controller} instance.
+     */
+    public ReaderGroupStateSynchronizer(String readerGroup, StateSynchronizer<ReaderGroupState> synchronizer, Controller controller) {
+        this.readerGroup = readerGroup;
+        this.synchronizer = synchronizer;
+        this.controller = controller;
+    }
+
+    /**
+     * Creates a new instance of ReaderGroupStateSynchronizer class.
+     *
+     * @param readerGroup Name of the {@link io.pravega.client.stream.ReaderGroup}.
+     * @param updateSerializer The serializer for {@link ReaderGroupState} updates.
+     * @param initSerializer The serializer for the initial {@link ReaderGroupState} update.
+     * @param config The {@link SynchronizerConfig}.
+     * @param clientFactory The ClientFactory instance.
+     * @param controller The {@link Controller} instance.
+     */
+    public ReaderGroupStateSynchronizer(String readerGroup,
+                                        Serializer<InitialUpdate<ReaderGroupState>> initSerializer,
+                                        Serializer<Update<ReaderGroupState>> updateSerializer,
+                                        SynchronizerConfig config,
+                                        SynchronizerClientFactory clientFactory,
+                                        Controller controller) {
+        this.readerGroup = readerGroup;
+        String streamName = getStreamForReaderGroup(readerGroup);
+        this.synchronizer = clientFactory.createStateSynchronizer(streamName, updateSerializer, initSerializer, config);
+        this.controller = controller;
+    }
+
+    @Override
+    public ReaderGroupState getState() {
+        return synchronizer.getState();
+    }
+
+    /**
+     * Method to fetch and apply all updates to the {@link ReaderGroupState} while making calls to the {@link Controller}
+     * to make sure the {@link ReaderGroupState} is consistent with its version on the {@link Controller).
+     */
+    @Override
+    public void fetchUpdates() {
+        synchronizer.fetchUpdates();
+        val clientState = getState();
+        val controllerState = getThrowingException(controller.getReaderGroup(readerGroup));
```

Comment:
There is no "RGState" stored on Controller only RGConfig.
Please check PR: https://github.com/pravega/pravega/pull/5400 for details.

Reply:
Once way to avoid making a controller API call on every fetchUpdates is to add a new boolean flag to RGState to indicate if SS was updated after controller update.
So inside fetchUpdates we can do:
```
fetchUpdates() {
// apply updates to RGState
if (ss.updatingConfig==true ) {
   controllerRGConfig = controller.getRG();
   if (ss.config.generation < controllerRGConfig.generation) {
      ss.updateStateUnconditionally(RGInit(controllerRGConfig with updatingConfig=false))
       }
  }
}
```